pkg/v1/remote/catalog.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+)
+
+type catalog struct {
+	Repos []string `json:"repositories"`
+}
+
+// GetCatalog calls /_catalog, returning the list of repositories on the registry
+func GetCatalog(target name.Registry, last string, n int, options ...Option) ([]string, error) {
```

Comment:
How do you feel about `GetCatalog` just doing pagination itself and returning all the values it can find? How many pages of results do we expect to get from this? Tens of thousands?

Reply:
That's why I was hesitant to implement it because I'm not sure of the scale.
I was planning on writing a helper for my own purposes that I can add later, but I just needed this at a minimum.

Reply:
The scale might be a concern, but for (most?) users who should expect only a couple pages, it sucks that this makes them handle pagination themselves.

I'd prefer to expose the pagination-handling version for the users that don't want to deal with it, and maybe if we hear that this method is a pain for whatever scaling reason, we can export the DIY method later. If we _never_ hear that we need to expose the guts, that's great.

Does that make any sense?

Reply:
If we're going to handle the pagination internally, we should probably make `GetCatalog` take a `context.Context` so that you can cancel it.

> it sucks that this makes them handle pagination themselves

This is basically why the `crane` package exists. IMO we should (within reason) expose all the complexity that the API itself has, then we can expose the happy path in `crane`.

Reply:
Another pattern I've seen that is nice, is to have the GetCatalog take a callback function that gets called per page. It allows users to abort early (via returning an error / cancelling the context), but it hides the pagination detail from the user.

Reply:
@ImJasonH thinks they're not idiomatic go. There's [this presentation](https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view) for inspiration.

I'm not sure if we could borrow something from that, but it also doesn't like callbacks (from [here](https://about.sourcegraph.com/go/gophercon-2018-rethinking-classical-concurrency-patterns)):

> Bryan briefly mentions asynchronous callbacks as something programmers from other certain languages sometimes try to use. But he notes that most Go programmers already know not to use them in Go, and moves on to talking about Futures and Producer–Consumer Queues.

Reply:
This is a synchronous callback - not a async one. The next page wouldn't be fetched until the callback returns.

Reply:
Ok, how do we feel about my new `crane.GetCatalog`?